sdk/eidolon_ai_sdk/agent/simple_agent.py

```diff
@@ -122,9 +214,48 @@ async def _act(self, action: ActionDefinition, process_id, **kwargs) -> AsyncIte
         env = Environment(undefined=StrictUndefined)
         text_message = UserTextCPUMessage(prompt=env.from_string(action.user_prompt).render(**body))
 
-        thread = await self.cpu.main_thread(process_id)
+        if execute_on_cpu:
+            cpu = self.cpu
+            for named_cpu in self.cpus:
+                if named_cpu.title == execute_on_cpu:
+                    cpu = named_cpu
+                    break
+        else:
+            cpu = self.cpu
+
+        thread = await cpu.main_thread(process_id)
         response = thread.stream_request(output_format=action.output_schema, prompts=[*image_messages, text_message])
 
         async for event in response:
             yield event
         yield AgentStateEvent(state=action.output_state)
+
+    async def _gen_title(self, action: ActionDefinition, process_id, **kwargs) -> AsyncIterable[StreamEvent]:
+        execute_on_cpu = None
+        request_body = to_jsonable_python(kwargs.get("body") or {})
+        if "execute_on_cpu" in request_body:
+            execute_on_cpu = request_body.pop("execute_on_cpu")
+
+        body = dict(datetime_iso=datetime.now().isoformat(), body=str(request_body))
+        if isinstance(request_body, dict):
+            body.update(request_body)
+
+        env = Environment(undefined=StrictUndefined)
+        text_message = UserTextCPUMessage(prompt=env.from_string(action.user_prompt).render(**body))
+
+        if execute_on_cpu:
+            cpu = self.cpu
+            for named_cpu in self.cpus:
+                if named_cpu.title == execute_on_cpu:
+                    cpu = named_cpu
+                    break
+        else:
+            cpu = self.cpu
+
+        title_message = UserTextCPUMessage(prompt=self.generate_title_prompt + text_message.prompt)
+        response = await (await cpu.new_thread(process_id)).run_request(prompts=[title_message])
+        process_obj = await ProcessDoc.find_one(query={"_id": process_id})
+        await process_obj.update(title=response)
+
+        yield StringOutputEvent(content=response)
+        yield AgentStateEvent(state=action.output_state)
```

Comment:
Nit: It would be nice here if we could return to the same state we were in before hand. perhaps we could throw previous state in the request context and grab it here?

Reply:
Done